shared/domain/src/main/java/com/ivy/domain/usecase/BalanceBuilder.kt

```diff
@@ -0,0 +1,63 @@
+package com.ivy.domain.usecase
+
+import com.ivy.data.model.primitive.AssetCode
+import com.ivy.data.model.primitive.NonZeroDouble
+import com.ivy.data.model.primitive.PositiveDouble
+
+class BalanceBuilder {
+
+    private val balance = mutableMapOf<AssetCode, NonZeroDouble>()
+
+    fun processDeposit(
+        incomes: Map<AssetCode, PositiveDouble>,
+        transferIn: Map<AssetCode, PositiveDouble>,
+    ) {
+        getAmount(incomes, transferIn).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + value.value)
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    fun processWithdrawal(
+        expenses: Map<AssetCode, PositiveDouble>,
+        transferOut: Map<AssetCode, PositiveDouble>,
+    ) {
+        getAmount(expenses, transferOut).forEach { (asset, value) ->
+            NonZeroDouble
+                .from((balance[asset]?.value ?: 0.0) + (-value.value))
+                .onRight { newValue ->
+                    balance[asset] = newValue
+                }
+        }
+    }
+
+    private fun getAmount(
+        firstMap: Map<AssetCode, PositiveDouble>,
+        secondMap: Map<AssetCode, PositiveDouble>,
+    ): Map<AssetCode, PositiveDouble> {
+        val amount = firstMap.toMutableMap()
+        combine(amount, secondMap)
+        return amount
+    }
+
+    private fun combine(
+        amount: MutableMap<AssetCode, PositiveDouble>,
+        secondMap: Map<AssetCode, PositiveDouble>
+    ) {
+        secondMap.forEach { (asset, value) ->
+            amount.merge(asset, value) { firstMap, secondMap ->
+                PositiveDouble
+                    .from(firstMap.value + secondMap.value)
+                    .fold(
+                        ifLeft = { PositiveDouble.unsafe(0.0) },
```

Comment:
`PositiveDouble.unsafe(0.0)` is 100% crash. The idea of PositiveDouble and other exact types is to guarantee some property (check the class impl). You should **never** use `unsafe` unless you can mathematically proof that the given property persists. Check [this docs](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/guidelines/Data-Modeling.md#data-modeling) - feel free to suggest improvements on top of it

You should use `PositiveDouble.from()`. Since `0.0` is not positive => that's not a positive double. It's a `NonNegativeDouble`